### PR TITLE
Override Stream.CopyTo in Streams that already override CopyToAsync

### PR DESCRIPTION
**Pull request by jamesqo**
Fixes https://github.com/dotnet/corefx/issues/12860

### CODE REVIEW

File: src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs [hunk 1]
```diff
@@ -1246,6 +1246,30 @@ public override void SetLength(long value)
             _stream.SetLength(value);
         }
 
+        public override void CopyTo(Stream destination, int bufferSize)
```
Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs]: The implementation here is largely copied from the `CopyToAsync` impl, with the inner `CopyToAsyncCore` method eliminated and written inline.

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 2]
```diff
@@ -679,6 +679,9 @@ public override Task WriteAsync(byte[] array, int offset, int count, Cancellatio
             }
         }
 
+        public override void CopyTo(Stream destination, int bufferSize) =>
+            CopyToAsync(destination, bufferSize).GetAwaiter().GetResult();
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: This is more likely to be a step backward than forward: it's now likely going to block a whole thread while relying on the copy to be done on thread pool threads.

File: src/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs [hunk 3]
```diff
@@ -420,6 +420,9 @@ public override void Write(ReadOnlySpan<byte> source)
 
         public override unsafe void WriteByte(byte value) =>
             Write(new ReadOnlySpan<byte>(&value, 1));
+        
+        public override void CopyTo(Stream destination, int bufferSize) =>
+            CopyToAsync(destination, bufferSize).GetAwaiter().GetResult();
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs]: Same as earlier; revert this.

File: src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs]: Where are the tests for this?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs [hunk 4]
```diff
@@ -85,6 +67,9 @@ public override Task FlushAsync(CancellationToken cancellationToken)
                 Task.CompletedTask;
         }
 
+        public override void CopyTo(Stream destination, int bufferSize) =>
+            CopyToAsync(destination, bufferSize).GetAwaiter().GetResult();
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs]: This file needs to compile for desktop as well, where this method isn't virtual.  And I don't think the value added here is significant enough to warrant splitting this out into a separate file... I'd just revert this file.

File: src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs]: @stephentoub I think I added them in a previous PR a while back, but I'll check to confirm. Admittedly I didn't think about tests until now.

Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs]: Never mind, https://github.com/dotnet/corefx/pull/9929/files was for `Stream.CopyTo`, not `BufferedStream.CopyTo`.

Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/IO/BufferedStream.cs]: @stephentoub I modified the existing `BufferedStream.CopyToAsync` tests to test `CopyTo` too.

File: src/Common/src/System/IO/DelegatingStream.cs [hunk 5]
```diff
@@ -172,6 +172,11 @@ public override void EndWrite(IAsyncResult asyncResult)
             _innerStream.EndWrite(asyncResult);
         }
 
+        public override void CopyTo(Stream destination, int bufferSize)
+        {
+            _innerStream.CopyTo(destination, bufferSize);
+        }
```
Comment (stephentoub) [inline src/Common/src/System/IO/DelegatingStream.cs]: Looks like this file is also built for netfx.  Do we ever actually expose this stream anywhere where CopyTo might be called on it?  If not, probably simplest to just revert this change, too.

Comment (jamesqo) [inline src/Common/src/System/IO/DelegatingStream.cs]: @stephentoub 

> Do we ever actually expose this stream anywhere where CopyTo might be called on it?

Yeah, we do. For example, `StreamContent.CreateContentReadStreamAsync` [returns a `ReadOnlyStream`](https://github.com/dotnet/corefx/blob/c351ba1f094871e4ae97f23f20f26648662cc4e6/src/System.Net.Http/src/System/Net/Http/StreamContent.cs) which inherits from this class, and the result of that can be passed out of `HttpContent.ReadAsStreamAsync` which is public.

What's the moniker I should put after #if here?

Comment (stephentoub) [inline src/Common/src/System/IO/DelegatingStream.cs]: It shouldn't be an #if but rather a partial class file with a .netcoreapp.cs extension.

Comment (stephentoub) [inline src/Common/src/System/IO/DelegatingStream.cs]: (That said, this use case also isn't particularly significant: the way you get such a stream is via async APIs: you're less likely at that point to switch to using sync ones, and if you do, you're likely already paying a perf penalty.)

Comment (jamesqo) [inline src/Common/src/System/IO/DelegatingStream.cs]: > the way you get such a stream is via async APIs: you're less likely at that point to switch to using sync ones

Really good point. Also, since this is in Common, if I move the override into a separate file people might forget to include it. I undid the changes here since it doesn't seem to be worth the trouble.

**Comment by jamesqo** [comment]
Test Linux x64 Release Build please

**Comment by stephentoub** [comment]
Test Linux x64 Release Build please

**Comment by stephentoub** [comment]
Thanks
